### PR TITLE
Lookup submision IDs by strategic metadata

### LINKED ISSUE #64: Support checking whether a submission currently exists in the queue for a particular piece of strategic metadata
In certain use-cases it can be useful to check whether a submission for that particular situation already exists. (And if it does, Producer code might for example decide to keep waiting for that submission to finish, or instead to cancel that one and start a new one).

Instead of requiring people to store that information separately from Opsqueue, we already have the strategic metadata available.

So the goal of this task is to introduce a new function to the ProducerClient, which you can give one (or multiple, if it isn't too much effort) key-values of strategic metadata, which returns a list of in-progress submissions matching that set of metadata.

For example, you could ask whether there are any submissions for `user_id: 42`.

### PR DESCRIPTION
**Pull request by jerbaroo**
This PR adds the `lookup_submission_ids_by_strategic_metadata` method to `ProducerClient`, which will return the submission IDs of in-progress subsmissions that match ALL of the provided strategic metadata key-value pairs.

```py
    def lookup_submission_ids_by_strategic_metadata(
        self, strategic_metadata: dict[str, int]
    ) -> list[SubmissionId]:
        """Attempts to find in-progress submissions where the strategic metadata
        of that submission includes all of the key-value pairs of the given
        'strategic_metadata'. A matching submission must include all of the
        given key-value pairs, but it may also contain other key-value pairs.

        Raises:
        - `TooManyMatchingSubmissionsError` if the lookup matches more
          submissions than the server's configured maximum. Narrow the query
          with more specific strategic metadata.
        - `InternalProducerClientError` if there is a low-level internal error.

        """
```

### CODE REVIEW

**Comment by ReinierMaas** [comment]
Open question shouldn't an empty `StrategicMetadataMap` return all submissions instead of raising an error on that?
Also this is an easy way to DDoS OpsQueue if the number of returned submissions is large.

File: libs/opsqueue_python/python/opsqueue/producer.py [hunk 1]
```diff
@@ -367,6 +371,25 @@ def lookup_submission_id_by_prefix(self, prefix: str) -> SubmissionId | None:
...
+        self, strategic_metadata: dict[str, int]
+    ) -> list[SubmissionId]:
+        """Attempts to find in-progress submissions where the strategic metadata
+        of that submission includes all of the key-value pairs of the given
+        'strategic_metadata'. A matching submission must include all of the
+        given key-value pairs, but it may also contain other key-value pairs.
+
+        Raises:
+        - `LookupIdsWithEmptyStrategicMetadataError` if the provided
+          'strategic_metadata' contained no key-value pairs to look for.
```
Comment (ReinierMaas) [inline libs/opsqueue_python/python/opsqueue/producer.py]: Can this potentially also raise:

```
- `InternalProducerClientError` if there is a low-level internal error.
```

File: libs/opsqueue_python/python/opsqueue/producer.py [hunk 2]
```diff
@@ -367,6 +371,25 @@ def lookup_submission_id_by_prefix(self, prefix: str) -> SubmissionId | None:
         """
         return self.inner.lookup_submission_id_by_prefix(prefix)
 
+    def lookup_submission_ids_by_strategic_metadata(
+        self, strategic_metadata: dict[str, int]
+    ) -> list[SubmissionId]:
```
Comment (ReinierMaas) [inline libs/opsqueue_python/python/opsqueue/producer.py:374]: I checked the other APIs on the producer and they either provide a single element or an iterator so that it can be lazily evaluated and doesn't need to be materialized in memory all at once.

File: opsqueue/src/producer/server.rs [hunk 3]
```diff
@@ -133,6 +139,16 @@ async fn lookup_submission_id_by_prefix(
     Ok(Json(submission_id))
 }
 
+async fn lookup_submission_ids_by_strategic_metadata(
+    State(state): State<ServerState>,
+    extract::Json(strategic_metadata): extract::Json<StrategicMetadataMap>,
+) -> Result<Json<Vec<SubmissionId>>, ServerError> {
```
Comment (ReinierMaas) [inline opsqueue/src/producer/server.rs]: Our `SubmissionId`s are close to `u64::max` in serialized size, which is reasonable for our random (snowflake) identifiers. This issues the following serialized to JSON response size:

* 1 (`[`) + elements * 20 + elements (`,|]`) characters

For ~1 million this is ~20MB. We can put that as an configurable upper bound to return to the client with an explicit error if it goes over that so that people can update the settings and return the larger response knowingly.

File: libs/opsqueue_python/python/opsqueue/producer.py (hunk 1 above)
Comment (ReinierMaas) [inline libs/opsqueue_python/python/opsqueue/producer.py]: I would change the interface to raise an explicit error if there are too many matching Submissions, we need that for DoS protection anyway, but allow an empty `strategic_metadata` set to be processed.

**Comment by jerbaroo** [comment]
> Open question shouldn't an empty StrategicMetadataMap return all submissions instead of raising an error on that?

It depends on the API contract we want to provide. An argument in favour of throwing an error is that if a caller is using an API to "search for submissions matching some strategic metadata" but they haven't provided any strategic metadata, then they possibly have an issue in their code that they should be alerted to. Also, the SQLite query will need a special case to handle this, more code for something we don't have evidence for that it is a usecase we want to support.

**Review by ReinierMaas — COMMENTED** [review]
I have a few comment and would like to point out that for the other queries we also assert the query plan. This allows us to spot potentially very bad execution behaviour.

File: libs/opsqueue_python/python/opsqueue/producer.py (hunk 1 above)
Comment (jerbaroo) [reply libs/opsqueue_python/python/opsqueue/producer.py]: Yes, good spot. The rust producer/client.rs code lifts network and decoding errors into this type. Documented in b557df0b1c2f2f57bf2db2c60da3be98d2456b95

Comment (jerbaroo) [reply libs/opsqueue_python/python/opsqueue/producer.py]: SGTM. Added support for empty `strategic_metadata` in 8057405. Will add the streaming / DoS protection in separate commits.

File: libs/opsqueue_python/python/opsqueue/producer.py (hunk 2 above)
Comment (jerbaroo) [reply libs/opsqueue_python/python/opsqueue/producer.py:374]: The other `ProducerClient` APIs which are returning iterators are returning _chunk_ iterators. These chunks are deterministically-addressable. Internally, the iterator is keeping track of a submission prefix, a current index, and a max index, then on each call to `next()` the iterator is performing a network request (`GET <object-storage-path><prefix>/<current-index>-out.bin`) to fetch the chunk.

This new API is different (not chunks, not querying object storage) from the existing APIs that return iterators, so would need some new wiring put in place to support streaming. But since we are only dealing with submission IDs, it's less of a memory concern. A pragmatic solution for now might be to just implement the configurable upper bound + add some metrics and keep an eye on it.

File: opsqueue/src/producer/server.rs (hunk 3 above)
Comment (jerbaroo) [reply opsqueue/src/producer/server.rs]: Configurable upper bound added in 54bc50bb974199124ffd4f53e136b1e4d4ad81ee with a roundtrip test in be5f6d2db02efddabab29228f63ef75d64c4be5a

Comment (jerbaroo) [reply opsqueue/src/producer/server.rs]: It might be a bit over-engineered with a newtype that ensures we don't overflow when adding (+1) when building the SQLite query, but.. types good right ? 😅

**Comment by jerbaroo** [comment]
> I have a few comment and would like to point out that for the other queries we also assert the query plan. This allows us to spot potentially very bad execution behaviour.

Query plan asserted in c6c9f04fa8c96f4041d6e82838b39c3b42d96dbf

File: justfile
Comment (ReinierMaas) [inline justfile:1]: I am not sure these timeouts are long enough on CI.

I had the same experience locally and made it configurable for local use:

https://github.com/channable/opsqueue/pull/122/changes#diff-deb9bb56fb122db0b605aa5b63f95a4665c905b18dd670e1fa6c877576a94ff1

File: libs/opsqueue_python/src/errors.rs [hunk 4]
```diff
@@ -146,6 +147,12 @@ impl From<CError<SubmissionNotFound>> for PyErr {
     }
 }
 
+impl From<CError<TooManyMatchingSubmissions>> for PyErr {
+    fn from(value: CError<TooManyMatchingSubmissions>) -> Self {
+        TooManyMatchingSubmissionsError::new_err(value.0 .0)
```
Comment (ReinierMaas) [inline libs/opsqueue_python/src/errors.rs]:
```suggestion
        TooManyMatchingSubmissionsError::new_err(value.0.0)
```

File: libs/opsqueue_python/python/opsqueue/producer.py [hunk 5]
```diff
@@ -38,10 +39,15 @@
     "SubmissionNotCancellable",
     "SubmissionNotCancellableError",
     "SubmissionNotFoundError",
+    "TooManyMatchingSubmissionsError",
     "ChunkFailed",
 ]
 
 
+class LookupIdsWithEmptyStrategicMetadataError(Exception):
+    pass
```
Comment (ReinierMaas) [inline libs/opsqueue_python/python/opsqueue/producer.py]: Is this error still produced?

File: opsqueue/src/common/mod.rs [hunk 6]
```diff
@@ -12,3 +12,50 @@ pub mod submission;
 /// consider hashing them and using that hash as MetaStateVal.
 pub type MetaStateVal = i64;
 pub type StrategicMetadataMap = FxHashMap<String, MetaStateVal>;
+
+/// Maximum number of submissions a lookup may return.
+/// Constructor checks: MaxSubmissions + 1 <= i64::MAX;
```
Comment (ReinierMaas) [inline opsqueue/src/common/mod.rs]: We should also prevent 0 from appearing, i.e. use `NonZeroU64` to store the value.

File: libs/opsqueue_python/tests/test_roundtrip.py [hunk 7]
```diff
@@ -508,3 +509,80 @@ def consume(x: int) -> int | None:
...
+        producer_client = ProducerClient(f"localhost:{opsqueue.port}", url)
+        inserted = 0
+        for _ in range(max_ + 1):
+            producer_client.insert_submission(
+                [1], chunk_size=1, strategic_metadata={"k": 1}
+            )
+            inserted += 1
+        with pytest.raises(TooManyMatchingSubmissionsError):
+            assert inserted == max_ + 1  # Make sure Exception wasn't raised too early.
+            producer_client.lookup_submission_ids_by_strategic_metadata({"k": 1})
```
Comment (ReinierMaas) [inline libs/opsqueue_python/tests/test_roundtrip.py]: We can also test the endpoint keeps functioning until the one that brings it over the edge, this prevents off-by-one errors:

```suggestion
        inserted = []
        for _ in range(max_ + 1):
            assert producer_client.lookup_submission_ids_by_strategic_metadata({"k": 1}) == inserted
            inserted.append(producer_client.insert_submission(
                [1], chunk_size=1, strategic_metadata={"k": 1}
            ))
        with pytest.raises(TooManyMatchingSubmissionsError):
            assert len(inserted) == max_ + 1  # Make sure Exception wasn't raised too early.
            producer_client.lookup_submission_ids_by_strategic_metadata({"k": 1})
```

File: opsqueue/src/common/submission.rs [hunk 8]
```diff
@@ -1075,6 +1146,38 @@ pub mod test {
...
+          submissions_metadata as s0
+          INNER JOIN submissions on submissions.id = s0.submission_id
+        WHERE
+          s0.metadata_key = ?
+          AND s0.metadata_value = ?
+        ORDER BY
+          s0.submission_id
+        LIMIT
+          ?
+        ");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs:1154]: This idea turned out to be so beautiful! It makes it easy to review the generated queries.

File: opsqueue/src/common/submission.rs [hunk 9]
```diff
@@ -550,6 +552,74 @@ pub mod db {
...
+    ) -> QueryBuilder<'_, Sqlite> {
+        let mut query_builder: QueryBuilder<Sqlite> =
+            QueryBuilder::new("SELECT s0.submission_id FROM submissions_metadata as s0");
+        // Inner join for each piece of strategic metadata.
+        for i in 1..strategic_metadata.len() {
+            query_builder.push(format!(
+                " JOIN submissions_metadata AS s{i} ON s{i}.submission_id = s0.submission_id"
+            ));
+        }
+        query_builder.push(" INNER JOIN submissions on submissions.id = s0.submission_id");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: This is to scope it to the `InProgress` submissions. We could make this `INNER JOIN` table configurable based on the requested `SubmissionState` enum.

File: opsqueue/src/common/submission.rs [hunk 10]
```diff
@@ -550,6 +552,74 @@ pub mod db {
...
+        }
+        let mut query_builder: QueryBuilder<Sqlite> =
+            lookup_ids_by_strategic_metadata_query(&strategic_metadata, limit);
+        let rows = query_builder.build().fetch_all(conn.get_inner()).await?;
+        error_if_too_many(
+            max_submissions,
+            rows.into_iter()
+                .map(|row| row.get("submission_id"))
+                .collect(),
+        )
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: We can simplify this a bit further to make it easier to follow:

```suggestion
        let ids = if strategic_metadata.is_empty() {
            // The main query to match on strategic_metadata will fail at run-time
            // if strategic_metadata is empty, so we handle the empty case here.
            query_scalar!(
                r#"SELECT id AS "id: SubmissionId" FROM submissions ORDER BY id LIMIT ?"#,
                limit
            )
            .fetch_all(conn.get_inner())
            .await?
        } else {
            let mut query_builder: QueryBuilder<Sqlite> =
                lookup_ids_by_strategic_metadata_query(&strategic_metadata, limit);
            query_builder.build_query_scalar().fetch_all(conn.get_inner()).await?
        };
        if ids.len() as u64 > u64::from(max_submissions) {
            // If 'max_submissions' results is exceeded return an explicit error.
            Err(E::R(TooManyMatchingSubmissions(u64::from(max_submissions))))
        } else {
            Ok(ids)
        }
```

---

Note that if we add a `SubmissionState` selector we need to switch out the `FROM submissions` with the correct `submission(_*)?` tables.

File: opsqueue/src/common/submission.rs [hunk 11]
```diff
@@ -550,6 +552,74 @@ pub mod db {
...
+    }
+
+    /// The query in 'lookup_ids_by_strategic_metadata', extracted for testing.
+    pub fn lookup_ids_by_strategic_metadata_query(
+        strategic_metadata: &StrategicMetadataMap,
+        limit: i64,
+    ) -> QueryBuilder<'_, Sqlite> {
+        let mut query_builder: QueryBuilder<Sqlite> =
+            QueryBuilder::new("SELECT s0.submission_id FROM submissions_metadata as s0");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: In the other query you added `AS "id: SubmissionId"` don't we need that here?

**Review by ReinierMaas — COMMENTED** [review]
I have a few code simplifications and one API generalisation.

File: opsqueue/src/common/submission.rs (hunk 11 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs]: The first query is verified at compile-time by sqlx, it looks at the DB schema and sees an `i64`, so we need to let sqlx know that it should read this column as `SubmissionId`. 

The second query is not verified at compile-time, we are just constructing a string and saying "trust me, it will return SubmissionIds at run-time"

File: justfile
Comment (jerbaroo) [reply justfile:1]: Great! Reverted the timeout in this PR in  9a85c035304a3d9f28ce6dfe9bc72ca366d2b066

File: libs/opsqueue_python/src/errors.rs (hunk 4 above)
Comment (jerbaroo) [reply libs/opsqueue_python/src/errors.rs]: Fixed in e09a1ea1e334997295ce7f116b7744e10966bc64

File: libs/opsqueue_python/python/opsqueue/producer.py (hunk 5 above)
Comment (jerbaroo) [reply libs/opsqueue_python/python/opsqueue/producer.py]: No, removed in 6d27f09a6ca98f6cbfdc09224f9570d17d435e96

File: opsqueue/src/common/mod.rs (hunk 6 above)
Comment (jerbaroo) [reply opsqueue/src/common/mod.rs]: Switched to `NonZero<u64>` in 441d2ca6e722d6ef9a5f4e06992dc3f92405d393

File: libs/opsqueue_python/tests/test_roundtrip.py (hunk 7 above)
Comment (jerbaroo) [reply libs/opsqueue_python/tests/test_roundtrip.py]: Good idea, accepted in e84bbc449c17b0b30a2c642ee1d65f688018173f

File: opsqueue/src/common/submission.rs (hunk 9 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs]: I like the idea, but would add it in a separate PR.

File: opsqueue/src/common/submission.rs (hunk 10 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs]: Nice simplification! Adopted in cbb1dd0948df0cc0771d57975268ab0170190777

File: opsqueue/src/common/submission.rs (hunk 9 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs]: Tracking in issue https://github.com/channable/opsqueue/issues/127

File: libs/opsqueue_python/tests/test_roundtrip.py [hunk 12]
```diff
@@ -508,3 +509,87 @@ def consume(x: int) -> int | None:
...
+            inserted.append(
+                producer_client.insert_submission(
+                    [1], chunk_size=1, strategic_metadata={"k": 1}
+                )
+            )
+        with pytest.raises(TooManyMatchingSubmissionsError):
+            assert (
+                len(inserted) == max_ + 1
+            )  # Make sure Exception wasn't raised too early.
+            producer_client.lookup_submission_ids_by_strategic_metadata({"k": 1})
```
Comment (ReinierMaas) [inline libs/opsqueue_python/tests/test_roundtrip.py]: I remember something about `pytest.raises` not always catching the error you expect and needing to have additional checks:

https://github.com/channable/imaginator/blob/75a752a630bacca36127824019d86e3aee152535/integration/frida/test_frida_extraction.py#L53-L59

File: opsqueue/src/common/submission.rs [hunk 13]
```diff
@@ -544,6 +546,63 @@ pub mod db {
...
+        max_submissions: MaxSubmissions,
+        mut conn: impl Connection,
+    ) -> Result<Vec<SubmissionId>, E<DatabaseError, TooManyMatchingSubmissions>> {
+        // MaxSubmissions provides us with the guarantee this won't overflow.
+        let limit = (u64::from(max_submissions) + 1) as i64;
+        // The main query to match on strategic_metadata will fail at run-time
+        // if strategic_metadata is empty, so we handle the empty case here.
+        let ids = if strategic_metadata.is_empty() {
+            query_scalar!(
+                r#"SELECT id AS "id: SubmissionId" FROM submissions ORDER by id LIMIT ?"#,
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]:
```suggestion
                r#"SELECT id AS "id: SubmissionId" FROM submissions ORDER BY id LIMIT ?"#,
```

File: opsqueue/src/common/submission.rs [hunk 14]
```diff
@@ -1069,6 +1129,38 @@ pub mod test {
         );
     }
 
+    #[sqlx::test]
+    pub async fn test_query_plan_lookup_by_strategic_metadata(db: sqlx::SqlitePool) {
+        let mut conn = db.acquire().await.unwrap();
+        let strategic_metadata: StrategicMetadataMap =
+            [("company_id".to_string(), 1), ("company_id".to_string(), 2)]
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: If you change one of the two stategic metadata keys to a different string this would cause us to see the recursion you wrote above.

File: opsqueue/src/common/submission.rs [hunk 15]
```diff
@@ -544,6 +546,63 @@ pub mod db {
...
+        }
+    }
+
+    /// The query in 'lookup_ids_by_strategic_metadata', extracted for testing.
+    pub fn lookup_ids_by_strategic_metadata_query(
+        strategic_metadata: &StrategicMetadataMap,
+        limit: i64,
+    ) -> QueryBuilder<'_, Sqlite> {
+        let mut query_builder: QueryBuilder<Sqlite> =
+            QueryBuilder::new("SELECT s0.submission_id FROM submissions_metadata as s0");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: Can we instead start of with the `submissions` table?

File: opsqueue/src/common/submission.rs [hunk 16]
```diff
@@ -544,6 +546,63 @@ pub mod db {
...
+        // Inner join for each piece of strategic metadata.
+        for i in 1..strategic_metadata.len() {
+            query_builder.push(format!(
+                " JOIN submissions_metadata AS s{i} ON s{i}.submission_id = s0.submission_id"
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: Then we can start from the `0` index here instead of skipping that.

File: opsqueue/src/common/submission.rs [hunk 17]
```diff
@@ -544,6 +546,63 @@ pub mod db {
...
+            ));
+        }
+        query_builder.push(" INNER JOIN submissions on submissions.id = s0.submission_id");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs]: And we can drop this line from the query altogether.

**Review by ReinierMaas — COMMENTED** [review]
I have some questions about the query we build-up. I don't think I fully follow without the recursive example working. And I expect that I hit on a small simplification.

File: libs/opsqueue_python/tests/test_roundtrip.py (hunk 12 above)
Comment (jerbaroo) [reply libs/opsqueue_python/tests/test_roundtrip.py]: added an additional assert on the type of exception and also checking it's inner value. All changes for this review round in 76d4ede

File: opsqueue/src/common/submission.rs (hunk 14 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs]: Fixed in 76d4ede and also applied the query simplification from your other comments where we select directly from `submissions` and avoid the additional join.

File: opsqueue/src/common/submission.rs [hunk 18]
```diff
@@ -544,6 +546,62 @@ pub mod db {
...
+            ));
+        }
+        for (i, (key, value)) in strategic_metadata.iter().enumerate() {
+            query_builder.push(if i == 0 { " WHERE" } else { " AND" });
+            query_builder.push(format!(" s{i}.metadata_key = "));
+            query_builder.push_bind(key.clone());
+            query_builder.push(format!(" AND s{i}.metadata_value = "));
+            query_builder.push_bind(*value);
+        }
+        query_builder.push(" ORDER BY s0.submission_id LIMIT ");
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs:598]: We can simplify this a bit further to a single for loop:

```suggestion
        // Inner join for each piece of strategic metadata.
        for (i, (key, value)) in strategic_metadata.iter().enumerate() {
            query_builder.push(format!(
                " INNER JOIN submissions_metadata AS s{i} ON s{i}.submission_id = submissions.id"
            ));
            query_builder.push(format!(" AND s{i}.metadata_key = "));
            query_builder.push_bind(key.clone());
            query_builder.push(format!(" AND s{i}.metadata_value = "));
            query_builder.push_bind(*value);
        }
        query_builder.push(" ORDER BY s0.submission_id LIMIT ");
```

File: opsqueue/src/common/submission.rs [hunk 19]
```diff
@@ -1069,6 +1128,42 @@ pub mod test {
...
+          JOIN submissions_metadata AS s1 ON s1.submission_id = submissions.id
+        WHERE
+          s0.metadata_key = ?
+          AND s0.metadata_value = ?
+          AND s1.metadata_key = ?
+          AND s1.metadata_value = ?
+        ORDER BY
+          s0.submission_id
+        LIMIT
+          ?
```
Comment (ReinierMaas) [inline opsqueue/src/common/submission.rs:1153]: Which changes the query to the following grouped statements:

```suggestion
        SELECT
          id
        FROM
          submissions
          INNER JOIN submissions_metadata AS s0 ON s0.submission_id = submissions.id
          AND s0.metadata_key = ?
          AND s0.metadata_value = ?
          INNER JOIN submissions_metadata AS s1 ON s1.submission_id = submissions.id
          AND s1.metadata_key = ?
          AND s1.metadata_value = ?
        ORDER BY
          s0.submission_id
        LIMIT
          ?
```

**Review by ReinierMaas — COMMENTED** [review]
LGTM! I was still thinking about a last small improvement to simplify the query generation code but I am not married to that.

**Review by ReinierMaas — APPROVED** [review]

**Comment by ReinierMaas** [comment]
The linter still failed on CI:

```diff
$ git diff
diff --git a/libs/opsqueue_python/tests/test_roundtrip.py b/libs/opsqueue_python/tests/test_roundtrip.py
index 2251c32..1f6ac50 100644
--- a/libs/opsqueue_python/tests/test_roundtrip.py
+++ b/libs/opsqueue_python/tests/test_roundtrip.py
@@ -582,13 +582,19 @@ def test_lookup_too_many_submission_ids_by_strategic_metadata() -> None:
         for _ in range(max_ + 1):
             assert (
                 inserted
-                == producer_client.lookup_submission_ids_by_strategic_metadata(strategic_metadata)
+                == producer_client.lookup_submission_ids_by_strategic_metadata(
+                    strategic_metadata
+                )
             )
             inserted.append(
-                producer_client.insert_submission([1], chunk_size=1, strategic_metadata=strategic_metadata)
+                producer_client.insert_submission(
+                    [1], chunk_size=1, strategic_metadata=strategic_metadata
+                )
             )
         with pytest.raises(TooManyMatchingSubmissionsError) as exc:
             assert len(inserted) == max_ + 1
-            producer_client.lookup_submission_ids_by_strategic_metadata(strategic_metadata)
+            producer_client.lookup_submission_ids_by_strategic_metadata(
+                strategic_metadata
+            )
         assert exc.type is TooManyMatchingSubmissionsError
         assert exc.value.max_submissions == max_

```

File: opsqueue/src/common/submission.rs (hunk 19 above)
Comment (jerbaroo) [reply opsqueue/src/common/submission.rs:1153]: Yay! Included in 23ed08c1b042804c394ddd0f5ae82f0f5494687f

**Comment by jerbaroo** [comment]
@OpsBotPrime merge and tag

**Comment by OpsBotPrime** [comment]
Your merge request has been denied, because merging on Fridays is not recommended. To override this behaviour use the command `merge and tag on Friday`.

**Comment by jerbaroo** [comment]
@OpsBotPrime merge and tag on friday

**Comment by OpsBotPrime** [comment]
Rebased as dcca94da91f71207e47eea2a045a58d3b5084336, waiting for CI …

**Comment by OpsBotPrime** [comment]
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/8bad3d4e-302e-4329-b9f7-c3345097bd22?pipeline_id=5c9744fb-19fc-4a40-9d11-9e596d7faf18) started.

**Comment by OpsBotPrime** [comment]
The [build failed :x:](https://channable.semaphoreci.com/workflows/8bad3d4e-302e-4329-b9f7-c3345097bd22?pipeline_id=5c9744fb-19fc-4a40-9d11-9e596d7faf18).

If this is the result of a flaky test, then tag me again with the `retry` command.  Otherwise, push a new commit and tag me again.

**Comment by jerbaroo** [comment]
@OpsBotPrime retry

**Comment by OpsBotPrime** [comment]
Your merge request has been denied, because merging on Fridays is not recommended. To override this behaviour use the command `retry on Friday`.

**Comment by jerbaroo** [comment]
@OpsBotPrime retry on friday

**Comment by OpsBotPrime** [comment]
Rebased as 931173e388b776fc3faaeb5f5184fe950b400826, waiting for CI …

**Comment by OpsBotPrime** [comment]
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/4d33bf51-0d59-4343-9636-22d6c4494b97?pipeline_id=17bc8cc3-c00a-4773-84ed-5a1a5716cd69) started.

**Comment by OpsBotPrime** [comment]
@jerbaroo I tagged your PR with [v57](https://github.com/channable/opsqueue/releases/tag/v57). Please wait for the build of 931173e388b776fc3faaeb5f5184fe950b400826 to pass and don't forget to deploy it!
